### PR TITLE
Add Observer render and inject sugar #387

### PR DESCRIPTION
**Pull request by Sunshine168**

### CODE REVIEW

**Comment by FredyC** [comment]
Just a hint, it might be easier to review this if you turn off Prettier :) (or perhaps eslint autofix?). Any kind of reformatting should be done in a separate PR in my opinion.

**Comment by Sunshine168** [comment]
Oh  i got a hobby which formate code anytime. i check package seem  `husky`  not be added, so precommit did not work well. 
i thought the rep should be format by same coding style and precommit should work, following this the code will be easier to review

hhhh,  i very appreciate your suggestion ,so i did it and will update typescript definitions later

**Comment by FredyC** [comment]
Yea, you are right, husky is not installed so `precommit` hook is not working. I've made another PR #407 to fix that. Once that get merged, you can simply rebase this and only relevant changes will be part of this PR.

**Comment by mweststrate** [comment]
@Sunshine168 Thanks for the PR! @FredyC's PR is merged, could you rebase?

PR looks good :). One remaining question; could you update the Readme as well describing this alternative syntax?

**Comment by FredyC** [comment]
Also, it's obvious that using `children` has a priority, but there should be test showing a use of both at once. And reflect this in docs as well.

**Comment by Sunshine168** [comment]
i will rebase later  , i am still in working

**Comment by Sunshine168** [comment]
done !

File: README.md [hunk 1]
```diff
@@ -82,7 +82,7 @@ It takes as children a single, argumentless function which should return exactly
 The rendering in the function will be tracked and automatically re-rendered when needed.
 This can come in handy when needing to pass render function to external components (for example the React Native listview), or if you
 dislike the `observer` decorator / function.
-
+And it can work with `Provider` use inject  and render property / or just takes as children, detail in Example2.(PS:using children has a priority than render , so dont use at the same time)
```
Comment (FredyC) [inline README.md]: I would split this, there are two concerns to cover - render prop and injecting. Also, I have a trouble understanding what it says. Let me rephrase this.

> In case you are a fan of render props, you can use that instead of children. Be advised, that you cannot use both approaches at once, children have a precedence.

_Here goes example showing render prop only_

> Observer can also inject the stores simply by passing a selector function.

_Second example with inject_

File: README.md [hunk 2]
```diff
@@ -105,6 +105,38 @@ React.render(<App person={person} />, document.body)
 person.name = "Mike" // will cause the Observer region to re-render
 ```
 
+Example2:
+
+```javascript
+class App extends React.Component {
+  render() {
+     return (
+          <Provider h="hello" w="world">
```
Comment (FredyC) [inline README.md]: Are you sure you can pass a string as a store prop? I am not aware that it would turn to observable automagically here. In that case, it's kinda misleading example.

File: src/observer.js [hunk 3]
```diff
@@ -349,12 +349,41 @@ function mixinLifecycleEvents(target) {
...
+        return component()
+    }
+    const InjectComponent = inject(observerInject)(component)
+    return <InjectComponent />
+})
 
 Observer.displayName = "Observer"
 
 Observer.propTypes = {
+    render: (propValue, key, componentName, location, propFullName) => {
```
Comment (FredyC) [inline src/observer.js]: DRYing this would be nice :)

File: README.md (hunk 1 above)
Comment (FredyC) [reply README.md]: Oh and with injecting it's worth to show an example that passing strings work same way as with `inject` HoC.

**Comment by Sunshine168** [comment]
@FredyC  Ohhhh  very thank you for your help    .tat   now will be better?

File: src/observer.js (hunk 3 above)
Comment (Sunshine168) [reply src/observer.js]: ennnn Any suggestion to diy  Observer.propTypes

File: README.md [hunk 4]
```diff
@@ -105,38 +103,63 @@ React.render(<App person={person} />, document.body)
...
-     )
-  }
-}
+
 const Comp = () => (
     <div>
         <Observer
-            inject={store => ({ h: store.h, w: store.w })}
-            render={props => <span>{`${props.h} ${props.w}`}</span>}
+            render={() => <span>hello world</span>}
```
Comment (FredyC) [inline README.md]: Hm, not very good example since there is nothing to observe :D Perhaps pass some variable in props to Comp and use that in render so it makes more sense.

File: README.md [hunk 5]
```diff
@@ -105,38 +103,63 @@ React.render(<App person={person} />, document.body)
 person.name = "Mike" // will cause the Observer region to re-render
 ```
 
-Example2:
+In case you are a fan of render props, you can use that instead of children. Be advised, that you cannot use both approaches at once, children have a precedence.
+
+Here goes example showing render prop only
```
Comment (FredyC) [inline README.md]: Well, I did not mean this line to be included there as well, that's why I used italic ;)

File: src/observer.js (hunk 3 above)
Comment (FredyC) [reply src/observer.js]: Well, if you look closely, it's the same function except for that first condition. Which doesn't make any sense to me. It means, that when checking `render` prop, you are telling it's ok if `children` prop is a function. That doesn't make any sense to me.

File: README.md [hunk 6]
```diff
@@ -105,6 +103,56 @@ React.render(<App person={person} />, document.body)
 person.name = "Mike" // will cause the Observer region to re-render
 ```
 
+In case you are a fan of render props, you can use that instead of children. Be advised, that you cannot use both approaches at once, children have a precedence.
+
+Here goes example showing render prop only
```
Comment (FredyC) [inline README.md]: This line is a bogus one, I did not mean to include it by using _italic text_ :)

File: README.md [hunk 7]
```diff
@@ -105,6 +103,56 @@ React.render(<App person={person} />, document.body)
...
+
+const person = observable({ name: "John" })
+
+React.render(<App person={person} />, document.body)
+person.name = "Mike" // will cause the Observer region to re-render
+```
+
+Observer can also inject the stores simply by passing a selector function.
+
+Example with inject
```
Comment (FredyC) [inline README.md]: Same for this line, it's redundant imo.

File: src/observer.js [hunk 8]
```diff
@@ -349,12 +349,41 @@ function mixinLifecycleEvents(target) {
 }
 
 // TODO: support injection somehow as well?
-export const Observer = observer(({ children }) => children())
+export const Observer = observer(({ children, inject: observerInject, render }) => {
+    const component = children || render
```
Comment (FredyC) [inline src/observer.js]: I am thinking if there shouldn't be some warning if both props are used. Perhaps that's what you had in mind with that propTypes check? Except you need to return message describing a problem. Returning undefined means it's valid.

Comment (Sunshine168) [reply src/observer.js]: ennn , thinging,  first i prefer show warning by propTypes if both props 
and coding like this 
```
const ObserverPropsCheck = (props,key,componentName,location,propFullName)=>{
    const extraKey = key === "children" ? "render" : "children"
    if(typeof propValue[key] === "function" && typeof propValue[extraKey] === "function" ){
            return new Error("Invalid prop,do not use children and render in the same time in`" + componentName )
    }
    
    if (typeof propValue[key] === "function" || typeof propValue[extraKey] === "function") {
        return
    }
    return new Error(
        "Invalid prop `" +
            propFullName +
            "` of type `" +
            typeof propValue[key] +
            "` supplied to" +
            " `" +
            componentName +
            "`, expected `function`."
    )
}

```

'>< but seem complexity     
or do same props check in Observer Component  instead in propType

Comment (FredyC) [reply src/observer.js]: Yea, checking with prop types like this is probably enough. I don't think it adds complexity, it's a friendly behavior toward user.

Comment (Sunshine168) [reply src/observer.js]: great !

File: README.md [hunk 9]
```diff
@@ -105,6 +103,53 @@ React.render(<App person={person} />, document.body)
...
+Example
+
+```javascript
+class App extends React.Component {
+  render() {
+     return (
+         <div>
+            {this.props.person.name}
+            <Observer
+              render={(props) => <div>{props.person.name}</div>}
```
Comment (Gvozd) [inline README.md]: `props` work without inejction?

Comment (FredyC) [reply README.md]: Good catch, this is an oversight. It should refer to `this.props` instead.

Comment (Sunshine168) [reply README.md]: Sorry ,a stupid fault  zzzzzz    should refer to this.props

**Comment by mweststrate** [comment]
Great work @Sunshine168 and thanks for the extensive reviews @FredyC and @Gvozd ! Will release tomorrow

**Comment by mvestergaard** [comment]
This is broken. Results in `Warning: Failed prop type: propValue is not defined` errors due to usage here https://github.com/mobxjs/mobx-react/blob/master/src/observer.js#L368

**Comment by FredyC** [comment]
@mvestergaard That's an unfortunate oversight. Could you send in the PR please? Some tests for this should be probably included as well.

**Comment by mweststrate** [comment]
Just released 4.4.1, that should fix the issue. Unclear yet why this passed the tests, so if somebody could add a test testing the prop types checking that would be appreciated :D
